File: vertx-elasticsearch/src/main/java/fr/myprysm/vertx/elasticsearch/action/search/aggregations/bucket/Range.java
/*
 * Copyright 2018 the original author or the original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.myprysm.vertx.elasticsearch.action.search.aggregations.bucket;

import fr.myprysm.vertx.elasticsearch.action.search.aggregations.Aggregation;
import io.vertx.codegen.annotations.DataObject;
import io.vertx.core.json.JsonObject;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.ToString;
import lombok.experimental.Accessors;

import java.util.Map;

@Data
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(callSuper = true)
@ToString(callSuper = true)
@Accessors(chain = true)
@DataObject(generateConverter = true)
public class Range extends Aggregation {

    private Map<String, RangeBucket> buckets;

    public Range(Range other) {
        super(other);
        buckets = other.buckets;
    }

    @SuppressWarnings("unchecked")
    public <T extends RangeBucket> T getBucketByKey(String term) {
        if (buckets == null) {
            return null;
        }

        return (T) buckets.get(term);

    }

    /**
     * Build a new <code>Range</code> from a <code>JsonObject</code>,
     * calling parent constructor.
     *
     * @param json the <code>JsonObject</code>
     */
    public Range(JsonObject json) {
        super(json);
        RangeConverter.fromJson(json, this);
    }

    /**
     * Transforms the <code>Range</code> into a <code>JsonObject</code>,
     * calling parent <code>toJson</code>.
     *
     * @return the <code>JsonObject</code>
     */
    public JsonObject toJson() {
        JsonObject json = super.toJson();
        RangeConverter.toJson(this, json);
        return json;
    }
}

File: vertx-elasticsearch/src/main/kotlin/fr/myprysm/vertx/elasticsearch/kotlin/action/search/aggregations/bucket/Range.kt
package fr.myprysm.vertx.elasticsearch.kotlin.action.search.aggregations.bucket

import fr.myprysm.vertx.elasticsearch.action.search.aggregations.bucket.Range
import fr.myprysm.vertx.elasticsearch.action.search.aggregations.bucket.RangeBucket

fun Range(
  buckets: Map<String, fr.myprysm.vertx.elasticsearch.action.search.aggregations.bucket.RangeBucket>? = null,
  data: io.vertx.core.json.JsonObject? = null,
  metaData: io.vertx.core.json.JsonObject? = null,
  name: String? = null,
  type: String? = null): Range = fr.myprysm.vertx.elasticsearch.action.search.aggregations.bucket.Range().apply {

  if (buckets != null) {
    this.setBuckets(buckets)
  }
  if (data != null) {
    this.setData(data)
  }
  if (metaData != null) {
    this.setMetaData(metaData)
  }
  if (name != null) {
    this.setName(name)
  }
  if (type != null) {
    this.setType(type)
  }
}

